### PR TITLE
migrate `prefer_foreach` from `traverseNodesInDFS()`

### PR DESCRIPTION
**Pull request by pq**
Surprisingly big gain in the general case (and we all know this can get quadratic in the pathological one). 🏁 

**BEFORE**

image

**AFTER**

image

/cc @srawlins @bwilkerson

### CODE REVIEW

**Review by srawlins — APPROVED** [review]
🎉

File: lib/src/rules/prefer_foreach.dart [hunk 1]
```diff
@@ -121,6 +116,29 @@ class _PreferForEachVisitor extends SimpleAstVisitor {
   }
 }
 
+class _ReferenceFinder extends GeneralizingAstVisitor {
```
Comment (bwilkerson) [inline lib/src/rules/prefer_foreach.dart:119]: This should use `UnifyingAstVisitor`.

File: lib/src/rules/prefer_foreach.dart [hunk 2]
```diff
@@ -121,6 +116,29 @@ class _PreferForEachVisitor extends SimpleAstVisitor {
...
+  bool found = false;
+  final LocalVariableElement? element;
+  _ReferenceFinder(this.element);
+
+  bool references(Expression target) {
+    if (target.canonicalElement == element) return true;
```
Comment (bwilkerson) [inline lib/src/rules/prefer_foreach.dart:125]: This test will be repeated in `visitNode`. If there isn't a performance benefit from not visiting the node (which there might well be if we expect this to return immediately often enough), then consider removing it.

Comment (pq) [reply lib/src/rules/prefer_foreach.dart:125]: It's a good point. I'm inclined to short-circuit visiting wherever possible but I haven't measured this...  I'll look a bit closer.  Thanks!

**Comment by pq** [comment]
(Landing as-is and will investigate the need for AST visit short-circuiting.)

**Pull request merged**